src/notebook/widget.ts

```diff
@@ -391,36 +409,81 @@ class NotebookWidget extends Widget {
     let layout = this._notebook.layout as PanelLayout;
     let constructor = this.constructor as typeof NotebookWidget;
     let factory = constructor.createCell;
-    let widget: Widget;
+    let widget: BaseCellWidget;
     switch(args.type) {
     case ListChangeType.Add:
-      layout.insertChild(args.newIndex, factory(args.newValue as ICellModel));
+      widget = factory(args.newValue as ICellModel);
+      layout.insertChild(args.newIndex, widget);
       break;
     case ListChangeType.Move:
       layout.insertChild(args.newIndex, layout.childAt(args.oldIndex));
       break;
     case ListChangeType.Remove:
-      widget = layout.childAt(args.oldIndex);
+      widget = layout.childAt(args.oldIndex) as BaseCellWidget;
       layout.removeChild(widget);
       widget.dispose();
       break;
     case ListChangeType.Replace:
       let oldValues = args.oldValue as ICellModel[];
       for (let i = args.oldIndex; i < oldValues.length; i++) {
-        widget = layout.childAt(args.oldIndex);
+        widget = layout.childAt(args.oldIndex) as BaseCellWidget;
         layout.removeChild(widget);
         widget.dispose();
       }
       let newValues = args.newValue as ICellModel[];
       for (let i = newValues.length; i < 0; i--) {
-        layout.insertChild(args.newIndex, factory(newValues[i]));
+        widget = factory(newValues[i])
+        layout.insertChild(args.newIndex, widget);
       }
       break;
     case ListChangeType.Set:
-      widget = layout.childAt(args.newIndex);
+      widget = layout.childAt(args.newIndex) as BaseCellWidget;
       layout.removeChild(widget);
       widget.dispose();
-      layout.insertChild(args.newIndex, factory(args.newValue as ICellModel));
+      widget = factory(args.newValue as ICellModel);
+      layout.insertChild(args.newIndex, widget);
+      break;
+    }
+    this.update();
+  }
+
+  /**
+   * Handle `update-request` messages sent to the widget.
+   */
+  protected onUpdateRequest(msg: Message): void {
+    // Set the appropriate classes on the cells.
+    let model = this.model;
+    let layout = this._notebook.layout as PanelLayout;
+    let widget = layout.childAt(model.activeCellIndex) as BaseCellWidget;
+    if (model.mode === 'edit') {
+      this.addClass(EDIT_CLASS);
+      this.removeClass(COMMAND_CLASS);
+      if (widget) {
```

Comment:
This was the focus of my comment about setting redundant classes on the cell widgets. I don't think we should set edit and command classes on the cells, only on the notebook widgets. Then the cell CSS can be `.EDIT_CLASS .cell_class` or `.COMMAND_CLASS .cell_class`.